### DISCUSSION TITLE
MorphToMany Accessing Blank Models?

### DISCUSSION THREAD

**Discussion by jjjrmy** [laravel/nova-issues#3747] [Q&A]
- Laravel Version: 8
- Nova Version: 3.31
- PHP Version: 8
- Database Driver & Version: MySQL

### Description:
If you add a Ray debug to the relationship method, the API request is returning blank models as shown below when accessing a Post resource.
```php
class Post extends Model {
  public function tags() {
     ray(request()->url(), $this)->label(__FUNCTION__);
     return $this->morphToMany(Tag::class, 'taggable');
  }
}
```
In my example this Post has 3 Tags, and it seem to be outputting 3 blank Post models in the debug output.
This is causing me problems and doesn't seem right.
![Ray Output](https://i.imgur.com/6dbMN4y.png)

Why is this happening and is there a way to prevent it?

### Detailed steps to reproduce the issue on a fresh Nova installation:
1.) Follow this structure for creating Laravel & Nova Models:
https://laravel.com/docs/8.x/eloquent-relationships#many-to-many-polymorphic-relations
2.) Create / Seed the Tags and Posts + Videos
3.) Navigate to a Post

**Comment by jjjrmy** [comment]
@crynobone can you please tell me what info is needed?

**Comment by crynobone** [comment]
All bug report should come with full reproducing code, not just code snippets.

**Comment by jjjrmy** [comment]
You can just copy the code from the Laravel documentation, but here it is...
**Migration**
```php
class CreateMorphsTable extends Migration
{
    public function up()
    {
        Schema::create('posts', function (Blueprint $table) {
            $table->id();
            $table->text('name');
            $table->timestamps();
        });
        Schema::create('videos', function (Blueprint $table) {
            $table->id();
            $table->text('name');
            $table->timestamps();
        });
        Schema::create('tags', function (Blueprint $table) {
            $table->id();
            $table->text('name');
            $table->timestamps();
        });
        Schema::create('taggables', function (Blueprint $table) {
            $table->id();
            $table->foreignId('tag_id');
            $table->morphs('taggable');
            $table->timestamps();
        });
    }
}
```

**Models**
```php
class Video extends Model
{
    public function tags()
    {
        return $this->morphToMany(Tag::class, 'taggable');
    }
}

class Post extends Model
{
    public function tags()
    {
        ray(request()->url(), $this)->label(__FUNCTION__);
        return $this->morphToMany(Tag::class, 'taggable');
    }
}

class Tag extends Model
{
    public function posts()
    {
        return $this->morphedByMany(Post::class, 'taggable');
    }

    public function videos()
    {
        return $this->morphedByMany(Video::class, 'taggable');
    }
}
```

**Nova Models**
```php
class Video extends Resource
{
    public function fields(Request $request)
    {
        return [
            ID::make(__('ID'), 'id')->sortable(),
            Text::make('Name'),
            MorphToMany::make('Tags'),
        ];
    }
}

class Post extends Resource
{
    public function fields(Request $request)
    {
        return [
            ID::make(__('ID'), 'id')->sortable(),
            Text::make('Name'),
            MorphToMany::make('Tags'),
        ];
    }
}

class Tag extends Resource
{
    public function fields(Request $request)
    {
        return [
            ID::make(__('ID'), 'id')->sortable(),
            Text::make('Name'),
        ];
    }
}
```

**Comment by jjjrmy** [comment]
I know it's doing something it shouldn't because if you access `Post::first()->tags` it will Ray dump just 1 instance of the Post model and no other blank models, but what ever is happening in the API Resource is calling those extra blank models.

**Comment by crynobone** [comment]
Calling blank model occurred in some API request and that is to be expected depending on what information needed for that particular API request.

I don't think there any bug here, as that is expected behaviour.

**Comment by jjjrmy** [comment]
@crynobone 
Can you please explain more because this doesn't seem to be expected behavior...

If you do `Post::find(1)->tags` in Tinkerwell, Ray will only dump one output of the Post model.
If you access `/nova/resources/posts/1` it calls the API `/nova-api/tags?...&viaResource=posts&viaResourceId=1&viaRelationship=tags&relationshipType=morphToMany`, Ray will dump as shown in the image above.
(In both cases 3 tags are attached to the Post.)

They are both accessing the same data, what ever is happening in that API endpoint I wouldn't call "to be expected" as that doesn't occur when doing it outside of Nova.

What is happening in the endpoint that is causing this?

**Comment by crynobone** [comment]
above are only your expectations, what we do is only `Tag::query()`, `new Tag()` or `Tag::find()`

**Reply by jjjrmy** [reply]
That makes sense but why would any of those return blank models of the Post?

**Comment by davidhemphill** [comment]
Depending on the setup, Nova sometimes spins up a resource that doesn't have a Model instance hydrated. For example, if the field is set up to not defer the query (e.g a `Select` field with its `options` directly querying the database like `Select::make('Type')->options(Type::all())` and not returned from a `Closure`), then Nova can sometimes access those relationships and they'll be empty. That's just one example. Nova does A LOT of this kind of thing.

What we need you to provide is the full reproducing code, along with descriptions of the problems you're facing. "This is causing me problems and doesn't seem right." is a little vague. Will be glad to turn this back into an issue once we get that. Thanks!

**Reply by jjjrmy** [reply]
The code I posted reproduces the issue I'm seeing..

In my attribute/relationship it calls relationships to other models which don't exist because the model is empty.
So is the solution to make a check if the model is empty before returning anything?

If you look at my example code, it's not using a Select field.
To me this seems like a Nova only issue/workaround?

**Reply by davidhemphill** [reply]
I'm just curious, wouldn't the field need to be a `MorphMany`, and not a `MorphToMany`?

**Reply by jjjrmy** [reply]
I can try that but I'm just copying the relationship setup as mentioned here: https://laravel.com/docs/8.x/eloquent-relationships#many-to-many-polymorphic-relations

Post -> morphToMany -> Tags
Video -> morphToMany -> Tags

Tag -> morphedByMany -> Posts/Videos

**Reply by jjjrmy** [reply]
Okay so I switched the relationship on Nova to `MorphMany` and kept the actual Model Relationship is still `MorphToMany`, and it seems to be working now with no issues of running additional blank models.

But I'm still curious as to why this is happening, to me it still seems like a bug.

**Reply by crynobone** [reply]
MorphToMany is correct usage since you have a pivot table `taggable`. You are not seeing empty model access there since it doesn't need to check for pivot as MorphMany doesn't have it.

**Comment by crynobone** [comment]
> Calling blank model occurred in some API request and that is to be expected depending on what information needed for that particular API request.

```php
    /**
     * Get all of the tags for the post.
     *
     * @return \Illuminate\Database\Eloquent\Relations\MorphToMany
     */
    public function tags()
    {
        ray(request()->fullUrl(), $this, collect(debug_backtrace())->transform(function ($stack) {
            if (!isset($stack['file']) || !isset($stack['line'])) {
                return null;
            }

            return $stack['file'].':'.$stack['line'];
        })->all())->label(__FUNCTION__);
        
        return $this->morphToMany(Tag::class, 'taggable')->withPivot('notes');
    }
```

Code above added to `Post` model on `nova-dusk-suite` which has a similar setup to the posted code.

![image](https://user-images.githubusercontent.com/172966/154451106-cd4b3a52-5ee2-40d1-956d-9fdf3358698f.png)

Line 707 on `ResolvesFields.php` class is where we call `Post::tags()` to get an instance of `Illuminate\Database\Eloquent\Relations\MorphToMany` (and not `Post` model) which allows us to retrieve the pivot accessor, we're not pulling any data for that purpose and that's the intended behavior
